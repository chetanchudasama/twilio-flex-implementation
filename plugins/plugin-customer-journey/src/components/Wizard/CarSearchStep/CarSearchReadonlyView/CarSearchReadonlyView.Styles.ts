import styled from "react-emotion";
import { defaultSnackbarColor } from "@common/components";

export const CarSearchReadonlyViewStyles = styled("div")({
  padding: "20px",
  margin: "-20px",
  backgroundColor: defaultSnackbarColor,
  ".item-value": {
    fontSize: "14px",
    padding: "8px 0px",
  },
  ".item-container": {
    padding: "7px",
  },
  ".edit-search-container": {
    textAlign: "right",
    paddingTop: "10px",
  },
  ".edit-icon": {
    fontSize: "18px",
    paddingRight: "5px",
  },
});
