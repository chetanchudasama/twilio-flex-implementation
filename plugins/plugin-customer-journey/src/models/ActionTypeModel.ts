export class ActionTypeModel {
  public id: number;

  public name: string;

  constructor() {
    this.id = -1;
    this.name = "";
  }
}
