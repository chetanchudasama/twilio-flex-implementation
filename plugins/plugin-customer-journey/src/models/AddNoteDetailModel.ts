export class AddNoteDetailModel {
  public applicationId: number;

  public content: string;

  public isImportant: boolean;

  public sendToBroker: boolean;

  constructor() {
    this.applicationId = -1;
    this.content = "";
    this.isImportant = false;
    this.sendToBroker = false;
  }
}
