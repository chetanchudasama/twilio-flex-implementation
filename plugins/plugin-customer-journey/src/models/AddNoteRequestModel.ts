import { Manager } from "@twilio/flex-ui";

export class AddNoteRequestModel {
  public comment: string;

  public applicationId: number;

  public userId: number;

  public isImportant: boolean;

  constructor() {
    const { agentId } =
      Manager.getInstance().store.getState().flex.worker.attributes;

    this.comment = "";
    this.applicationId = -1;
    this.userId = agentId || 0;
    this.isImportant = false;
  }
}
