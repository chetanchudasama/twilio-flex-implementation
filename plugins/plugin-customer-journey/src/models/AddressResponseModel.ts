export class AddressResponseModel {
  public CompanyName: string;

  public SubBuilding: string;

  public BuildingName: string;

  public BuildingNumber: string;

  public Street: string;

  public Locality: string;

  public Town: string;

  public Country: string;

  public Postcode: string;

  public AddressLine: string;

  public MaskedPostcode: string;

  public Key: string;

  public GridEast: string;

  public GridNorth: string;

  public Latitude: number;

  public Longitude: number;

  constructor() {
    this.CompanyName = "";
    this.SubBuilding = "";
    this.BuildingName = "";
    this.BuildingNumber = "";
    this.Street = "";
    this.Locality = "";
    this.Town = "";
    this.Country = "";
    this.Postcode = "";
    this.AddressLine = "";
    this.MaskedPostcode = "";
    this.Key = "";
    this.GridEast = "";
    this.GridNorth = "";
    this.Latitude = -1;
    this.Longitude = -1;
  }
}
