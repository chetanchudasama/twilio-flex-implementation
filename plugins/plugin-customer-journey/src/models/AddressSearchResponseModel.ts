export class AddressSearchResponseModel {
  public Address: string;

  public Key: string;

  constructor() {
    this.Address = "";
    this.Key = "";
  }
}
