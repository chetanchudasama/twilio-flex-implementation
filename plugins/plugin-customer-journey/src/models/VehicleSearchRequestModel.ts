import { Shared } from "@common/components";
import { VehicleSortingOptions } from "../common/enum";
import { RangeFilterModel } from "./RangeFilterModel";

export class VehicleSearchRequestModel {
  public makeIds: number[];

  public modelIds: number[];

  public engineMin: number | null;

  public engineMax: number | null;

  public ageRange: RangeFilterModel | null;

  public mileageRange: RangeFilterModel | null;

  public priceMax: number | null;

  public priceMonthly: number | null;

  public bodyIds: number[];

  public fuelIds: number[];

  public colourIds: number[];

  public transId: number | null;

  public distance: number | null;

  public keywords: string | null;

  public includeDeliveryDealer: boolean;

  public isFlexible: boolean;

  public page: number | null;

  public pageSize: number | null;

  public sort: number | null;

  public postcode: string;

  public dealerId: number | null;

  constructor() {
    this.makeIds = [];
    this.modelIds = [];
    this.engineMin = null;
    this.engineMax = null;
    this.ageRange = null;
    this.mileageRange = null;
    this.priceMax = null;
    this.priceMonthly = null;
    this.bodyIds = [];
    this.colourIds = [];
    this.fuelIds = [];
    this.transId = null;
    this.distance = null;
    this.keywords = "";
    this.includeDeliveryDealer = false;
    this.isFlexible = false;
    this.page = Shared.defaultPage;
    this.pageSize = Number(
      process.env.REACT_APP_CAR_SEARCH_PAGE_SIZE || Shared.defaultPageSize
    );
    this.sort = VehicleSortingOptions.PriceDescending;
    this.postcode = "";
    this.dealerId = null;
  }
}
