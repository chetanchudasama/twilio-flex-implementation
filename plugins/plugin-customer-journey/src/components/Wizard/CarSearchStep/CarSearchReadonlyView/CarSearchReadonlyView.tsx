import React from "react";
import { Grid, InputLabel, Button, Icon } from "@material-ui/core";
import { Shared, ItemModel, VehicleModelData } from "@common/components";
import { v4 } from "uuid";
import { CarSearchReadonlyViewStyles } from "./CarSearchReadonlyView.Styles";
import { VehicleSearchDropdownData } from "../../../../models/VehicleSearchDropdownData";
import { VehicleSearchRequestModel } from "../../../../models/VehicleSearchRequestModel";

export interface CarSearchReadonlyViewProps {
  vehicleSearchDropdownData: VehicleSearchDropdownData;
  filters: VehicleSearchRequestModel;
  term: number | null;
  partExchangeValue: number | null;
  partExchangeSettlement: number | null;
  deposit: number | null;
  handleEditSearch: () => void;
  vrm: string;
  dealerInfo: string;
}

const CarSearchReadonlyView: React.FC<CarSearchReadonlyViewProps> = ({
  vehicleSearchDropdownData,
  filters,
  term,
  partExchangeValue,
  partExchangeSettlement,
  deposit,
  handleEditSearch,
  vrm,
  dealerInfo,
}) => {
  const getSelectedItems = (
    itemList: VehicleModelData[] | ItemModel[],
    selectedItems: string | number | (string | number)[] | undefined
  ) => {
    const list: string[] = [];
    (selectedItems as number[]).map((value) =>
      list.push(itemList.find((item) => item.id === value)?.name ?? "")
    );
    return list.join(", ");
  };

  const modelItems = () => {
    const modelList: VehicleModelData[] = [];
    vehicleSearchDropdownData.makesModels
      .filter((item) => filters.makeIds?.includes(item.id))
      .map((model) => model.models.map((el) => modelList.push(el)));
    return modelList;
  };

  const filterValue = (): string => {
    if (filters?.engineMax && filters?.engineMin) {
      return `${filters?.engineMin} - ${filters?.engineMax}`;
    }

    if (filters?.engineMin && !filters?.engineMax) {
      return `greater than ${filters?.engineMin}`;
    }

    if (!filters?.engineMin && filters?.engineMax) {
      return `less than ${filters?.engineMax}`;
    }

    return "";
  };

  const searchItems = [
    {
      label: "Car make",
      value: getSelectedItems(
        vehicleSearchDropdownData.makesModels,
        filters?.makeIds
      ),
      hasValue: filters?.makeIds?.length > 0,
    },
    {
      label: "Car model",
      value: getSelectedItems(modelItems(), filters?.modelIds),
      hasValue: filters?.modelIds?.length > 0,
    },
    {
      label: "Distance",
      value: `Within ${filters?.distance} miles`,
      hasValue: !!(filters?.distance && filters?.distance !== -1),
    },
    {
      label: "Dealer",
      value: dealerInfo,
      hasValue: !!(!vrm && dealerInfo),
    },
    {
      label: "Keywords",
      value: filters?.keywords,
      hasValue: !!filters?.keywords,
    },
    {
      label: "Include dealers who deliver",
      value: "Yes",
      hasValue: filters?.includeDeliveryDealer === true,
    },
    {
      label: "Body type",
      value: getSelectedItems(vehicleSearchDropdownData.body, filters?.bodyIds),
      hasValue: filters?.bodyIds.length > 0,
    },
    {
      label: "Maximum mileage",
      value: filters?.mileageRange?.max,
      hasValue: !!filters?.mileageRange?.max,
    },
    {
      label: "Transmission",
      value: vehicleSearchDropdownData.trans
        .filter((item) => item.id === filters?.transId)
        .map((transmission) => transmission.name),
      hasValue: !!(filters?.transId && filters?.transId !== -1),
    },
    {
      label: "Maximum age",
      value: filters?.ageRange?.max,
      hasValue: !!filters?.ageRange?.max,
    },
    {
      label: "Fuel type",
      value: getSelectedItems(vehicleSearchDropdownData.fuel, filters?.fuelIds),
      hasValue: filters?.fuelIds.length > 0,
    },
    {
      label: "Colour",
      value: getSelectedItems(
        vehicleSearchDropdownData.colour,
        filters?.colourIds
      ),
      hasValue: filters?.colourIds.length > 0,
    },
    {
      label: "Engine size",
      value: filterValue(),
      hasValue: !!(filters?.engineMin || filters?.engineMax),
    },
  ];

  return (
    <CarSearchReadonlyViewStyles>
      <Grid container>
        {filters && !vrm && (
          <>
            <Grid item xs={5}>
              {searchItems
                .filter((item) => item.hasValue)
                .map((searchItem) => (
                  <Grid item xs={12} className="item-container" key={v4()}>
                    <InputLabel>{searchItem.label}</InputLabel>
                    <div className="item-value">{searchItem.value}</div>
                  </Grid>
                ))}
            </Grid>
            <Grid item xs={1} />
            <Grid item xs={6}>
              <Grid container item xs={12}>
                <Grid item xs={6}>
                  {filters?.priceMax && filters?.priceMax > 0 ? (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Total Price</InputLabel>
                      <div className="item-value">
                        {Shared.getFormattedCurrencyValue(filters.priceMax)}
                      </div>
                    </Grid>
                  ) : null}
                  {filters?.priceMonthly && filters?.priceMonthly > 0 ? (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Monthly Budget</InputLabel>
                      <div className="item-value">
                        {Shared.getFormattedCurrencyValue(filters.priceMonthly)}
                      </div>
                    </Grid>
                  ) : null}
                  {partExchangeValue && (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Part Ex. Value</InputLabel>
                      <div className="item-value">
                        {Shared.getFormattedCurrencyValue(partExchangeValue)}
                      </div>
                    </Grid>
                  )}
                  {deposit && (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Deposit</InputLabel>
                      <div className="item-value">
                        {Shared.getFormattedCurrencyValue(deposit)}
                      </div>
                    </Grid>
                  )}
                </Grid>
                <Grid item xs={6}>
                  {term && (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Term</InputLabel>
                      <div className="item-value">
                        {Shared.termList
                          .filter((item) => item.id === term)
                          .map((termItem) => termItem.name)}
                      </div>
                    </Grid>
                  )}
                  {filters.isFlexible === true && (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Flexible budget (+/-&pound;50)</InputLabel>
                      <div className="item-value">Yes</div>
                    </Grid>
                  )}
                  {partExchangeSettlement && (
                    <Grid item xs={12} className="item-container">
                      <InputLabel>Part Ex. Settlement</InputLabel>
                      <div className="item-value">
                        {Shared.getFormattedCurrencyValue(
                          partExchangeSettlement
                        )}
                      </div>
                    </Grid>
                  )}
                </Grid>
              </Grid>
            </Grid>
          </>
        )}
        <Grid item xs={5}>
          {vrm && (
            <Grid item xs={12} className="item-container">
              <InputLabel>VRM</InputLabel>
              <div className="item-value">{vrm}</div>
            </Grid>
          )}
        </Grid>
        <Grid item xs={4}>
          {dealerInfo && vrm && (
            <Grid item xs={12} className="item-container">
              <InputLabel>Dealer</InputLabel>
              <div className="item-value">{dealerInfo}</div>
            </Grid>
          )}
        </Grid>
        <Grid item xs={3} className="edit-search-container">
          <Button onClick={handleEditSearch} size="large">
            <Icon className="edit-icon" fontSize="small">
              edit
            </Icon>
            Edit Search
          </Button>
        </Grid>
      </Grid>
    </CarSearchReadonlyViewStyles>
  );
};

export default CarSearchReadonlyView;
