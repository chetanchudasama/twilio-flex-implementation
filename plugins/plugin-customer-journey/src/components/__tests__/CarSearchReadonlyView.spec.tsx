import React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";
import { ItemModel } from "@common/components";
import CarSearchReadonlyView, {
  CarSearchReadonlyViewProps,
} from "../Wizard/CarSearchStep/CarSearchReadonlyView/CarSearchReadonlyView";
import { VehicleSearchRequestModel } from "../../models/VehicleSearchRequestModel";

const items: ItemModel[] = [
  { id: 1, name: "test1" },
  { id: 2, name: "test2" },
  { id: 3, name: "test3" },
];

const filters = Object.assign(new VehicleSearchRequestModel(), {
  makeIds: [1, 2],
  modelIds: [1, 2],
  bodyIds: [1, 2],
  colourIds: [1, 2],
  fuelIds: [1, 2],
});

const props: CarSearchReadonlyViewProps = {
  vehicleSearchDropdownData: {
    makesModels: [
      {
        id: 1,
        name: "test1",
        models: [
          { id: 1, name: "test1", count: 1 },
          { id: 2, name: "test2", count: 2 },
        ],
      },
      {
        id: 2,
        name: "test2",
        models: [
          { id: 1, name: "test1", count: 1 },
          { id: 2, name: "test2", count: 2 },
        ],
      },
      {
        id: 3,
        name: "test3",
        models: [
          { id: 1, name: "test1", count: 1 },
          { id: 2, name: "test2", count: 2 },
        ],
      },
    ],
    body: items,
    colour: items,
    fuel: items,
    trans: items,
    distance: [0, 10, 20],
  },
  filters,
  term: 48,
  partExchangeValue: 220,
  partExchangeSettlement: 400,
  deposit: 300,
  vrm: "",
  dealerInfo: "",
  handleEditSearch: jest.fn(() => {}),
};

describe("CarSearchReadonlyView", () => {
  it("renders correctly", async () => {
    const wrapper = create(<CarSearchReadonlyView {...props} />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("renders correctly as per filters prop value", async () => {
    const wrapper = create(<CarSearchReadonlyView {...props} />).root;

    expect(wrapper.findAllByProps({ className: "item-value" })).toHaveLength(9);
  });

  it("when new search item is added, should rendered in UI", async () => {
    props.filters.keywords = "test-value";
    const wrapper = create(<CarSearchReadonlyView {...props} />).root;

    expect(wrapper.findAllByProps({ className: "item-value" })).toHaveLength(
      10
    );
  });

  it("when filter contains vrm value, should rendered in UI", async () => {
    props.vrm = "test";
    const wrapper = create(<CarSearchReadonlyView {...props} />).root;

    expect(wrapper.findAllByProps({ className: "item-value" })).toHaveLength(1);
  });

  it("when edit search button is clicked, handleEditSearch method should be called", () => {
    const wrapper = mount(<CarSearchReadonlyView {...props} />);

    const editButton = wrapper
      .find(".edit-search-container")
      .find("button")
      .at(0);
    editButton.simulate("click");
    wrapper.update();

    expect(props.handleEditSearch).toBeCalledTimes(1);
  });
});
