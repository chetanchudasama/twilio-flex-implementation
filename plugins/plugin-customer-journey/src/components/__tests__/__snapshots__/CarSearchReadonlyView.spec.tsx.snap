// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarSearchReadonlyView renders correctly 1`] = `
<div
  className="css-17w034h"
>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-5-34"
    >
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
      >
        <label
          className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
        >
          Car make
        </label>
        <div
          className="item-value"
        >
          test1, test2
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
      >
        <label
          className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
        >
          Car model
        </label>
        <div
          className="item-value"
        >
          test1, test2
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
      >
        <label
          className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
        >
          Body type
        </label>
        <div
          className="item-value"
        >
          test1, test2
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
      >
        <label
          className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
        >
          Fuel type
        </label>
        <div
          className="item-value"
        >
          test1, test2
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
      >
        <label
          className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
        >
          Colour
        </label>
        <div
          className="item-value"
        >
          test1, test2
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-6-35"
    >
      <div
        className="MuiGrid-container-1 MuiGrid-item-2 MuiGrid-grid-xs-12-41"
      >
        <div
          className="MuiGrid-item-2 MuiGrid-grid-xs-6-35"
        >
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
          >
            <label
              className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
            >
              Part Ex. Value
            </label>
            <div
              className="item-value"
            >
              £220
            </div>
          </div>
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
          >
            <label
              className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
            >
              Deposit
            </label>
            <div
              className="item-value"
            >
              £300
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-item-2 MuiGrid-grid-xs-6-35"
        >
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
          >
            <label
              className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
            >
              Term
            </label>
            <div
              className="item-value"
            >
              48 months
            </div>
          </div>
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 item-container"
          >
            <label
              className="MuiFormLabel-root-109 MuiInputLabel-root-98 MuiInputLabel-animated-106"
            >
              Part Ex. Settlement
            </label>
            <div
              className="item-value"
            >
              £400
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-5-34"
    >
      
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-4-33"
    >
      
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-3-32 edit-search-container"
    >
      <button
        className="MuiButtonBase-root-142 MuiButton-root-116 MuiButton-text-118 MuiButton-flat-121 MuiButton-sizeLarge-140"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-117"
        >
          <span
            aria-hidden="true"
            className="material-icons MuiIcon-root-145 MuiIcon-fontSizeSmall-152 edit-icon"
          >
            edit
          </span>
          Edit Search
        </span>
        <span
          className="MuiTouchRipple-root-154"
        />
      </button>
    </div>
  </div>
</div>
`;
